### PR TITLE
Enable via UI rather than code

### LINKED ISSUE #37: Enable via UI rather than code
Enabling and disabling via code is cumbersome and not very efficient or convenient for our needs. Would be much more ideal if this could be controlled from the admin dashboard.

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #37 

**Description**

Allow users to activate the maintenance mode via customizer in addition to the existing activation option via adding `define( 'VIP_MAINTENANCE_MODE', true )` to functions.php (WordPress.com) resp. to vip-config.php (VIP Go).

**Steps to test**

1. Check out [the featured branch](https://github.com/Automattic/maintenance-mode-wp/tree/update/%2337)
2. Install and activate the plugin on a testing site
3. Go to the frontend of the testing site to ensure that the maintenance mode is not active
4. Go to `WP Admin → Appearance → Customize → Maintenance Mode` and activate the maintenance mode
5. Go to the frontend of the testing site to ensure that the maintenance mode is active

**Screenshots**
<table>
<tr>
<td>

mm-01
</td>
<td>

mm-03
</td>
<td>

mm-04
</td>
</tr>
</table>

### CODE REVIEW

File: maintenance-mode.php [hunk 1]
```diff
@@ -3,56 +3,154 @@
  * Plugin Name: Maintenance Mode
  * Plugin URI: https://vip.wordpress.com/plugins/maintenance-mode/
  * Description: Shut down your site for a little while and do some maintenance on it!
- * Author: Automattic / WordPress.com VIP
+ * Author: Automattic / WordPress VIP
```
Comment (GaryJones) [inline maintenance-mode.php]: How is changing the author name relevant to adding a UI?

File: maintenance-mode.php [hunk 2]
```diff
@@ -3,56 +3,154 @@
...
  * Author URI: https://vip.wordpress.com
- * Version: 0.2.2
+ * Version: 0.3.0
```
Comment (GaryJones) [inline maintenance-mode.php]: Keep version bumps to separate PR's, ideally in a release branch, along with other changes like docs improvements, changelogs, etc.

While _you_ may think this is the only new set of changes that warrant a new release, there could be other changes (like tidying Plugin URI, updating Author, other small fixes, etc.) that could go into a new release. 

Likewise, the addition of a UI could be all it needs to consider this to be a stable release, and so a 1.0.0 version would be warranted. (I personally don't think it would be, but I've not found any discussion around this.)

File: maintenance-mode.php [hunk 3]
```diff
@@ -3,56 +3,154 @@
...
  * License: GPLv2
  * Text Domain: maintenance-mode
  * Domain Path: /languages
+ */
+
+/**
+ * Avoid direct plugin access
```
Comment (GaryJones) [inline maintenance-mode.php]: Why is there a DocBlock (`/**`) for a simple function _call_?

A simple `// Avoid direct plugin access.` would be sufficient here. The `@since` isn't important, especially as this DocBlock wouldn't be extractable anywhere.

Also, you'd added `0.2.3`, but then you've tried to make this release go from 0.2.2 to 0.3.0 anyway ;-)

File: maintenance-mode.php [hunk 4]
```diff
@@ -3,56 +3,154 @@
...
 
-// Stops the execution early if the VIP_MAINTENANCE_MODE constant is not set to `true`.
-if ( !defined( 'VIP_MAINTENANCE_MODE' ) || true !== VIP_MAINTENANCE_MODE ) {
-	add_action( 'admin_notices', 'vip_maintenance_mode_admin_notice__constant_not_set' );
-	return;
+/**
+ * Localize plugin
+ *
+ * @since 0.1.1
+ * @uses load_plugin_textdomain
```
Comment (GaryJones) [inline maintenance-mode.php]: Don't bother with these `@uses` annotations - they aren't really used, and easily become out of date or incomplete since they look at the implementation of a function (the insides), rather than be _about_ the function.

File: maintenance-mode.php [hunk 5]
```diff
@@ -3,56 +3,154 @@
...
+ */
+function vip_maintenance_mode_load_plugin_textdomain() {
```
Comment (GaryJones) [inline maintenance-mode.php]: What was the advantage in moving this function to the top of the file? How does that help with adding a UI?

File: maintenance-mode.php [hunk 6]
```diff
@@ -3,56 +3,154 @@
...
+	load_plugin_textdomain( 'maintenance-mode', FALSE, basename( dirname( __FILE__ ) ) . '/languages/' );
 }
+add_action( 'plugins_loaded', 'vip_maintenance_mode_load_plugin_textdomain' );
 
 /**
- * Displays a warning in the admin to inform users about the VIP_MAINTENANCE_MODE constant
+ * Add settings link on plugin page
```
Comment (GaryJones) [inline maintenance-mode.php]: As per https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/, all Summary lines in a DocBlock should end with a full stop. Fix for all instances.

File: maintenance-mode.php [hunk 7]
```diff
@@ -3,56 +3,154 @@
...
  *
- * Only users who can activate plugins will see the warning.
- *
- * @since 0.1.1
+ * @since 1.0.0
```
Comment (GaryJones) [inline maintenance-mode.php]: As per https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods there should be a one-line-space after the `@since` tag line. Fix for all instances.

Also, how can this be a `1.0.0` when the plugin version was supposedly only going to be `0.3.0`?

File: maintenance-mode.php [hunk 8]
```diff
@@ -3,56 +3,154 @@
...
+ * @param array $links The original array with customizer links.
+ * @return array $links The updated array with customizer links.
```
Comment (GaryJones) [inline maintenance-mode.php]: Swap `customizer` for `Customizer` since that's the name (proper noun) of the feature.

Comment (GaryJones) [inline maintenance-mode.php]: `@return` tags shouldn't include a variable name, since functions return values, not variables.

File: maintenance-mode.php [hunk 9]
```diff
@@ -3,56 +3,154 @@
...
-	$message = __( 'Maintenance Mode won\'t work until you set the VIP_MAINTENANCE_MODE constant to <code>true</code>.', 'maintenance-mode' );
-	printf( '<div class="%1$s"><p>%2$s</p></div>', esc_attr( $class ), wp_kses( $message, array( 'code' => array() ) ) );
+/**
+ * Checks to see if the maintenance mode is activated
+ * 
+ * @since 0.2.3
+ * @return bool Return is maintenance mode is activated.
+ */
+function vip_maintenance_mode_is_activated() {
+	return get_option( 'vip_maintenance_mode_enable' ) || defined( 'VIP_MAINTENANCE_MODE' ) || current_user_can_bypass_vip_maintenance_mode();
```
Comment (GaryJones) [inline maintenance-mode.php:43]: Let's say I used your refactored code and enabled MM with `define( 'VIP_MAINTENANCE_MODE', true );` - all good, the site is restricted.

Now, it comes time to launch, so I change `true` to `false`, and ... the site is still restricted.

Why?

Comment (GaryJones) [reply maintenance-mode.php:43]: Looking at this function and the function below - if I have the default capability (`edit_posts`), and so `current_user_can_bypass_vip_maintenance_mode()` returns true, then the logic in this line will say that MM should be activated, and thus restrict me.

Yet, if I was a subscriber, and don't have that capability, then the function below will return false, and so it won't restrict me.

You should consider breaking this up into more logical steps:

```php
if ( current_user_can_bypass_vip_maintenance_mode() ) {
	return false; // If they have the right capabilities, don't enable MM for them, regardless of setting or constant.
}

if ( get_option( 'vip_maintenance_mode_enable' ) || defined( 'VIP_MAINTENANCE_MODE' ) ) {
	return true; // MM is indicated to be enabled. 
}

return false;
```

...which could then potentially be slightly simplified to:

```php
if ( current_user_can_bypass_vip_maintenance_mode() ) {
	return false; // If they have the right capabilities, don't enable MM for them, regardless of setting or constant.
}

// MM is indicated to be enabled. 
return get_option( 'vip_maintenance_mode_enable' ) || defined( 'VIP_MAINTENANCE_MODE' );
```

File: maintenance-mode.php [hunk 10]
```diff
@@ -3,56 +3,154 @@
...
+	global $wp_admin_bar;
+
+	if ( ! vip_maintenance_mode_is_activated() ) {
+		return;
+	}
+
+	$wp_admin_bar->add_menu( array(
+		'id'     => 'maintenance-mode',
+		'parent' => 'top-secondary',
+		'title'  => apply_filters( 'vip_maintenance_mode_admin_bar_title', __( 'Under maintenance', 'maintenance-mode' ) ),
```
Comment (GaryJones) [inline maintenance-mode.php:75]: Could this status be linked through to the Customizer page, of that's where it had been enabled, and a user has sufficient rights to view that page / make that change?

File: maintenance-mode.php [hunk 11]
```diff
@@ -3,56 +3,154 @@
...
+		color: #ECEFF4 !important;
+		font-weight:500;
+	}';
+	wp_add_inline_style( 'admin-bar', $styles );
+}
+add_action( 'wp_enqueue_scripts', 'vip_maintenance_mode_admin_scripts' );
+add_action( 'admin_enqueue_scripts', 'vip_maintenance_mode_admin_scripts' );
+
+/**
+ * Enhance customizer
```
Comment (GaryJones) [inline maintenance-mode.php]: This Summary is too broad to describe what's really happening here.

File: maintenance-mode.php [hunk 12]
```diff
@@ -3,56 +3,154 @@
...
+ *
+ * @since 0.2.3
+ * @param WP_Customize_Manager $wp_customize The instance of the WP_Customize_Manager class.
+ */
+function vip_maintenance_mode_register_customize( $wp_customize ) {
```
Comment (GaryJones) [inline maintenance-mode.php]: `WP_Customize_Manager` could be added as a function argument type declaration, just to add a little bit more protection that `->add_*()` methods will exist (or more specifically, that the fatal error would be a bit more specific if `$wp_customize` is a different class).

File: maintenance-mode.php [hunk 13]
```diff
@@ -3,56 +3,154 @@
...
+
+/**
+ * Sanitize customizer checkbox input
+ *
+ * @since 0.2.3
+ * @param bool $input The boolean to sanitize.
+ * @return bool The sanitized boolean.
+ */
+function vip_maintenance_mode_sanitize_checkbox( $input ) {
+	return isset( $input ) ? true : false;
```
Comment (GaryJones) [inline maintenance-mode.php]: What value (if any) does the submission contain if the checkbox is empty? If it's `null`, then all good. If it's `''` or `0` or `false` (which is what you have as the `default`), then this will return an incorrect result. Consider:

```shell
$ php -a
php > $input = null; echo isset( $input );
php > $input = ''; echo isset( $input );
1
php > $input = 0; echo isset( $input );
1
php > $input = false; echo isset( $input );
1
```

See https://github.com/ahmadawais/WPCustomize/blob/bdf7b9cc5fe8242630c9ac8e1de65132129d6ada/customizer/customizer-sanitization.php#L28-L40 for a safer approach to consider.

File: maintenance-mode.php [hunk 14]
```diff
@@ -61,7 +159,11 @@ function current_user_can_bypass_vip_maintenance_mode() {
  * @since 0.1.1
  */
 function vip_maintenance_mode_template_redirect() {
-	if ( current_user_can_bypass_vip_maintenance_mode() ) {
+	if ( is_customize_preview() ) {
```
Comment (GaryJones) [inline maintenance-mode.php:153]: This seems like it's a specific fix to a specific problem. While it may have been discovered in creating the setting in the Customizer, I think this should be pulled into its own PR, since we likely wouldn't want this fix reverted if the rest of the UI feature had to be reverted (I find that to be a good way of thinking about what code should be split out into other PRs).

File: maintenance-mode.php [hunk 15]
```diff
@@ -106,7 +208,15 @@ function vip_maintenance_mode_template_redirect() {
 }
 add_action( 'template_redirect', 'vip_maintenance_mode_template_redirect' );
 
+/**
+ * Restrict REST API access
+ *
+ * @since 0.2.0
+ * @param  mixed $result The original REST API result.
+ * @return WP_Error|null|true WP_Error if authentication error, null if authentication method wasn't used, true if authentication succeeded.
+ */
```
Comment (GaryJones) [inline maintenance-mode.php]: It's good to add docs, but this should be a separate PR.

File: readme.txt [hunk 16]
```diff
@@ -1,9 +1,9 @@
 === Maintenance Mode ===
```
Comment (GaryJones) [inline readme.txt]: Keep changelog updates separate from the PR - they can be put together just before a release is made.

Also see https://johnblackbourn.com/readme-md-plugin-on-wordpress-org/ - in a separate PR you could scrap the `readme.txt` file completely and move it all over to `readme.md`

File: readme.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
+# Maintenance Mode
+
+[![Build Status](https://api.travis-ci.com/Automattic/maintenance-mode.svg?branch=master)](https://api.travis-ci.com/Automattic/maintenance-mode)
+[![GPLv2 License](https://img.shields.io/github/license/Automattic/maintenance-mode.svg)](https://www.gnu.org/licenses/gpl.html)
```
Comment (GaryJones) [inline readme.md]: Use a version-specific links, since this gpl.html is for V3.

https://www.gnu.org/licenses/old-licenses/gpl-2.0.html is for V2.

**Review by GaryJones — CHANGES_REQUESTED** [review]
The addition of the UI setting is generally going to be welcomed. 

I think this PR does too much though; it can be hard to resist sneaking in extra fixes, but that's where self-discipline and Doing Things Right comes in - what is the minimum number of line changes that could be made to add this new feature in a secure and performant way?

File: maintenance-mode.php (hunk 1 above)
Comment (nielslange) [reply maintenance-mode.php]: It's not relevant and had been removed.

File: maintenance-mode.php (hunk 2 above)
Comment (nielslange) [reply maintenance-mode.php]: I see what you mean. I haven't thought about that yet and it makes a lot of sense to keep version bumps separate.

File: maintenance-mode.php (hunk 3 above)
Comment (nielslange) [reply maintenance-mode.php]: > Why is there a DocBlock (/**) for a simple function call?
I guess I tried to be uber-correct. I changed the multiline comment into a single line comment.

> Also, you'd added 0.2.3, but then you've tried to make this release go from 0.2.2 to 0.3.0 anyway ;-)
Oh. That is clearly wrong and had been fixed.

File: maintenance-mode.php (hunk 4 above)
Comment (nielslange) [reply maintenance-mode.php]: To my defence, I need to say that the `@uses` annotations hasn't been added by me, but, of course, I removed it.

File: maintenance-mode.php (hunk 5 above)
Comment (nielslange) [reply maintenance-mode.php]: I moved it up as I structure my personal plugins that way. To me, loading the textdomain is similar to the `<head>` sections within an HTML document, while the business logic of the plugin is placed in the `<body>` section. That said, this PR does not benefit from this change and I reverted it.

File: maintenance-mode.php (hunk 6 above)
Comment (nielslange) [reply maintenance-mode.php]: I treated this part as the title of a multiline comment and, therefore, I haven't added a full stop. However, I know understand that this is not the comment title, but the comment summary and I added full stops for all instances.

File: maintenance-mode.php (hunk 7 above)
Comment (nielslange) [reply maintenance-mode.php]: I've added the space for all instances and corrected the wrong version number.

File: maintenance-mode.php (hunk 8 above)
Comment (nielslange) [reply maintenance-mode.php]: Done ...

Comment (nielslange) [reply maintenance-mode.php]: ... and done. 😀

File: maintenance-mode.php (hunk 9 above)
Comment (nielslange) [reply maintenance-mode.php:43]: This logic is indeed error-prone, and I need to look closer into this one.

File: maintenance-mode.php (hunk 10 above)
Comment (nielslange) [reply maintenance-mode.php:75]: This is also existing code and I didn't want to adjust it this might be confusing for existing users.

File: maintenance-mode.php (hunk 11 above)
Comment (nielslange) [reply maintenance-mode.php]: This is also existing code, but I changed the message from `Styles the admin bar notice` to `Show admin bar notice when the maintenance mode had been activated.`.

File: maintenance-mode.php (hunk 12 above)
Comment (nielslange) [reply maintenance-mode.php]: I've added the type declaration and added `Requires PHP: 7.4` to the plugin header, to indicate which PHP version is required to run this plugin.

File: maintenance-mode.php (hunk 13 above)
Comment (nielslange) [reply maintenance-mode.php]: In this case, my check definitely delivered the wrong result. Thanks for pointing this out.

File: readme.md (hunk 17 above)
Comment (nielslange) [reply readme.md]: Good catch. I missed that one.

File: readme.txt (hunk 16 above)
Comment (nielslange) [reply readme.txt]: That was new to me and I moved the content from `readme.txt` to `readme.md` and removed `readme.txt`.

File: maintenance-mode.php (hunk 15 above)
Comment (nielslange) [reply maintenance-mode.php]: The documentation had been removed accordingly.

File: maintenance-mode.php (hunk 14 above)
Comment (nielslange) [reply maintenance-mode.php:153]: Moving `is_customize_preview()` into a separate PR would no longer allow a user to deactivate the maintenance mode once it had been activated. The only way to deactivate the maintenance mode then would be to deactivate either the plugin or to alter the `vip_maintenance_mode_enable` setting within the options table. Therefore, I don't think that removing it would be in the interest of the user, unless I'm missing something here. 🤔

**Review by nielslange — COMMENTED** [review]
@GaryJones Except of one point, I addressed all your feedback. That one point is regarding the logic to distinguish if MM is active or not. I'll need to look into this topic in detail.

File: maintenance-mode.php (hunk 3 above)
Comment (GaryJones) [reply maintenance-mode.php]: > I guess I tried to be uber-correct. I changed the multiline comment into a single line comment.

`/*` blocks and `/**` blocks serve different purposes. The first is a multi-line comment, the latter is a DocBlock.

File: maintenance-mode.php (hunk 11 above)
Comment (GaryJones) [reply maintenance-mode.php]: I wonder why this existing code is coming up in the diff then.

Could be worth moving these maintenance items (fixing existing DocBlocks, etc.) to a new PR, since most of them are going to be related to the new feature of enabling MM via the UI.

File: maintenance-mode.php (hunk 12 above)
Comment (GaryJones) [reply maintenance-mode.php]: What version of PHP is needed to add a class type declaration?

Hint: It's not PHP 7.4 ;-)

**Comment by GaryJones** [comment]
@nielslange I've unresolved a few comment threads where I've followed up.

File: maintenance-mode.php (hunk 3 above)
Comment (nielslange) [reply maintenance-mode.php]: Thanks for pointing this out.

**Comment by nielslange** [comment]
Closing this PR given the low number of active installations.
